### DISCUSSION TITLE
Need to transfer field values from a matrix into a supertable

### DISCUSSION THREAD

**Discussion by gbowne-quickbase** [craftcms/cms#13430] [Q&A] [answered]
We've got a complex matrix set up for our components, and early on we'd add fields to each component, but then started using supertables to contain them. I'd love to be able to move values into their supertables easily so authors don't have to. Here's an example with all the unnecessary stuff hidden

Matrix Field called `mainComponents` with 30+ items in it. Here's two:
- Hero
  - `sectionConfig` (supertable)
    - `backgroundColor` (dropdown)
    - `externalSpacing` (dropdown)
  - `backgroundColor` (field on the matrix for Hero)
  - `externalSpacing`  (field on the matrix for Hero)
- TextSpan  
  - `sectionConfig` (supertable)
    - `backgroundColor`
    - `externalSpacing`
  - `backgroundColor` (field on the matrix for Text Span)
  - `externalSpacing`  (field on the matrix for Text Span)

Since a component's background color might be set to `"blue"` we can't switch code over to just look at the `sectionConfig.backgroundColor` value which might still have the default value of `"white"` and we don't want authors to have to look through hundreds of pages.

**Comment by brandonkelly** [comment, marked as answer]
This can be accomplished using the [resave/matrix-blocks](https://craftcms.com/docs/3.x/console-commands.html#resave-matrix-blocks) command:

```sh
php craft resave/matrix-blocks \
  --field mainComponents \
  --type "hero" \
  --set sectionConfig \
  --to "fn(\$block) => ['new1' => ['type' => 1, 'fields' => ['backgroundColor' => \$block->backgroundColor, 'externalSpacing' => \$block->externalSpacing]]]"
```

Update `'type' => 1` with the actual Super Table block type ID. You can find it via a hidden input at the beginning of an input’s table row, within an entry edit page:

```html
<input type="hidden" name="fields[mainComponents][blocks][2149][fields][sectionConfig][blocks][2162][type]" value="1">
```

Repeat the command for each Matrix block type, replacing `hero` with the Matrix block type handle, and updating `'type' => 1` with the correct Super Table block type ID.

**Comment by gbowne-quickbase** [comment]
This is working great for us! One question: what is the `['new1' =>` in the last line of the block. Is that just a nonsense value to assign to the new data, or does it have meaning?

**Reply by brandonkelly** [reply]
It tells Matrix to interpret the data as a new block element. It could be `new` followed by any number.

**Comment by gbowne-quickbase** [comment]
After a few rounds of wondering why this wasn't working, I re-read all your instructions. So I'll put them here in case someone else needs the help.
When editing the `field` within the `Settings->Fields` portion of the CP, the `type` on a particular SuperTable variable is actually the id within the `matrixblocks` table and that made sense since it's part of that, but we need to go an actual instance of the page and look at the the `type` which is totally different. These come from the `supertableblocks` table, and would be a little harder to find directly, so definitely look for an instance of the matrix block to find what you're using.
